docs/setup/options.md
Configuration Options
=====================

Visit the [Configuration File](configure.md) section for information about
how to use these configuration options.

Workspace
---------

> This is the location where DTrack stores data. This includes saved recordings,
> audio clips, trained models, temporary data, etc.
>
> !!! option "Default Value: `"_workspace/"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | workspace              | DTRACK\_WORKSPACE         |


Workspace Keep Temp
-------------------

> Keep temporary data after extracted for review.
>
> !!! option "Default Value: `false`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | boolean | keep\_temp             | DTRACK\_KEEP\_TEMP        |

Record Audio Device
-------------------

> The device identifier used to record audio.
>
> !!! option "Default Value: `"plughw"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | audio\_device          | RECORD\_AUDIO\_DEVICE     |
>
> - **List audio devices:** `ffmpeg -loglevel warning -sources alsa`
>     + Sample Output:
>
>             Auto-detected sources for alsa:
>             null [Discard all samples (playback) or generate zero samples (capture)]
>             hw:CARD=Snowflake,DEV=0 [Direct hardware device without any conversions]
>             plughw:CARD=Snowflake,DEV=0 [Hardware device with all software conversions]
>             default:CARD=Snowflake [Default Audio Device]
>             sysdefault:CARD=Snowflake [Default Audio Device]
>             front:CARD=Snowflake,DEV=0 [Front output / input]
>             dsnoop:CARD=Snowflake,DEV=0 [Direct sample snooping device]
>
>     + Look for devices with: "<u>software conversions</u>"
>     + It is likely worth taking time to test available options.

Record Audio Options
--------------------

> Additional ffmpeg options used for audio capture.
>
> !!! option "Default Value: `[ "-f", "alsa" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | audio\_options         | RECORD\_AUDIO\_OPTIONS    |

Record Video Device
-------------------

> The device identifier used to record video.
>
> !!! option "Default Value: `"/dev/video0"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | video\_device          | RECORD\_VIDEO\_DEVICE     |
>
> - **List video devices:** `ls -1 /dev/video*`
>     + Sample Output:
>
>             /dev/video0
>             /dev/video1
>             /dev/video2
>             /dev/video3
>
> - **List capabilities:** `ffmpeg -hide_banner -f v4l2 -list_formats all -i <DEVICE>`
>     + Sample Output:
>
>             [v4l2 @ 0x5612] Compressed:       mjpeg: Motion-JPEG : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Compressed:        h264:       H.264 : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Compressed: Unsupported:        HEVC : 640x360 640x480 960x540 1024x576 1280x720 1920x1080 2560x1440 3840x2160
>             [v4l2 @ 0x5612] Raw       :     yuyv422:  YUYV 4:2:2 : 640x360 640x480 960x540 1024x576 1280x720 1920x1080
>             [in#0 @ 0x5612] Error opening input: Immediate exit requested
>             Error opening input file /dev/video2.
>
>     + This output shows a camera capable of recording 4K video.

Record Video Options
--------------------

> Additional ffmpeg options used for video capture.
>
> !!! option "Default Value: `[ "-f", "v4l2", "-framerate", "15" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | video\_options         | RECORD\_VIDEO\_OPTIONS    |

Record Video Timestamp
----------------------

> Long string that defines the datestamp that is added on top of recorded video.
>
> The default value uses `Monospace Bold`, set to a bright `Red, 48-Point` font.
> This font is provided by the `fonts-freefont-ttf` package,
>
> This filter is applied to every frame, so it is possible to add `hflip,vflip,`
> to the start of the string, which will result in an image that is flipped both
> horizontally and vertically--perfect for an upside-down camera.
>
> !!! option "Default Value: `"drawtext=fontfile=/usr/share/fonts/truetype/freefont/FreeMonoBold.ttf:text=%{localtime}:fontcolor=red@0.9:x=7:y=7:fontsize=48"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | video\_timestamp       | RECORD\_VIDEO\_TIMESTAMP  |

Record Video Advanced
---------------------

> Video creation arguments that ultimately control the final recorded video quality.
>
> A sensible range for 1080p video is between `3M` and `8M`, with `3.5M` being the
> observed upper limit on a Raspberry Pi 5--due to lack of hardware **en**coder.
>
> !!! option "Default Value: `[ "libx264", "-crf", "23", "-preset", "fast", "-tune", "zerolatency", "-maxrate", "3M", "-bufsize", "24M" ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | video\_advanced        | RECORD\_VIDEO\_ADVANCED   |

Record Inspect Models
---------------------

> List of tags used for training, and trained models used for automatic detection.
>
> !!! option "Default Value: `[ ]`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | list    | inspect\_models        | RECORD\_INSPECT\_MODELS   |

Record Inspect Backlog
----------------------

> Defines the maximum number of "segments" that can back up in each queue
> before newly recorded audio is discarded from queue.
>
> !!! option "Default Value: `5`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | number  | inspect\_backlog       | RECORD\_INSPECT\_BACKLOG  |

Record Duration
---------------

> Length of time to use for each recording.
>
> - Very short recordings will suffer inadequate compression and are harder to review.
> - Very long recordings will be much larger and more difficult to submit.
>
> !!! option "Default Value: `"00:10:00"`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | string  | record\_duration       | RECORD\_DURATION          |

Train Target
------------

> Success rate of model (as percent correct) required before training is
> considered a success.
>
> !!! option "Default Value: `0.95`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | number  | train\_target          | TRAIN\_TARGET             |

Train Rate
----------

> One of the "big three" macro adjustments in machine learning.
>
> !!! option "Default Value: `0.001`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | number  | train\_rate            | TRAIN\_RATE               |

Train Momentum
--------------

> One of the "big three" macro adjustments in machine learning.
>
> !!! option "Default Value: `0.9`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | number  | train\_momentum        | TRAIN\_MOMENTUM           |

Train Dropout
-------------

> One of the "big three" macro adjustments in machine learning.
>
> !!! option "Default Value: `0.2`"
>     | Type    | Configuration Variable | Environment Variable      |
>     | ------- | ---------------------- | ------------------------- |
>     | number  | train\_dropout         | TRAIN\_DROPOUT            |


docs/setup/configure.md
Configuration File
==================

DTrack has defaults that are designed to function adequately on a Raspberry
Pi 5, although these make very poor monitoring devices. If your [Selected
Hardware](hardware.md) provides a hardware video **en**conder, then it is
important to modify **record_save_options** (below).


config.json
-----------

Configuration is stored in a **JSON** file, called **config.json**.

**Important JSON Tips:**

1\. Format:
```
{
  "OPTION": VALUE,
  "OPTION": VALUE,
  "OPTION": VALUE
}
```

2\. All **VALUE**s are one of:

- **string**: ``"anythin inside of quotes"``
- **list**: ``["like", "a", "string", "broken", "into", "parts"]``
- **number**: ``5`` or ``3.1415``
- **boolean**: ``true`` for Yes or ``false`` for No

3\. Every **VALUE** must have a trailing comma (``,``), except the last must not.

Test-Driven Defaults
--------------------

Defaults were selected based on tests using [lowest-recommended hardware](hardware.md).

- [Encoder Documentation](https://trac.ffmpeg.org/wiki/Encode/H.264) (Upstream Defaults)

| Value                 | Purpose                                                            |
| --------------------- | ------------------------------------------------------------------ |
| ``-t <time>``         | Must be set on **each** input device                               |
| ``-tune zerolatency`` | This is **required** for software encoding on low-end cpu          |
| ``-bufsize 64M``      | Very large buffer to helps avoid processing spikes                 |
| ``-crf 23``           | Default is best; 25 reduces size by 30%, but 24 creates movement   |
| ``-maxrate 3M``       | Hard quality limit, based on **NO hardware encoding**              |
|                       | * 1080p recordings will see quality improvement up to about ``7M`` |
|                       | * Larger value creates larger files and XBUF for CPU-only encoding |
| ``-preset fast``      | Yield fewest XRUN errors                                           |
| ``-framerate 15``     | Fast enough to catch most movement                                 |

**Baseline Command:**

```
ffmpeg -y -loglevel warning -nostdin -nostats \
    -t 00:10:00 -f alsa -i plughw \
    -t 00:10:00 -f v4l2 -i /dev/video0 \
    -map 0:a -c:a pcm_s16le -ar 48000 -ac 1 -f wav - \
    -filter_complex [1:v]...[dtstamp] -map 0:a -map [dtstamp] \
    -c:a pcm_s16le -ar 48000 -ac 1 -c:v libx264 baseline.mkv \
    >/dev/null
```


src/state/config.go
// Application Configuration
//
// ALL VARIABLES:
// - Define Runtime variables and types:	Application_Configuration
// - Default value of Runtime variables:	Load_Configuration
// - Map JSON Settings to Runtime variables:	Application_Configuration
// - Map Environment vars to Runtime variables:	Environment_Configation_Map
// - Help text showing JSON/EnvVar/Default:	Show_Help
package state

import (
	// DTrack
	"dtrack/log"

	// Standard
	"encoding/json"
	"fmt"
	"os"
	"reflect"
	"strconv"
)

// Master object holding loaded configuration data
var Runtime Application_Configuration

// Map json configuration to Runtime
// Defaults set in load_config()
type Application_Configuration struct {
	Workspace              string   `json:"workspace"`
	Workspace_Keep_Temp    bool     `json:"workspace_keep_temp"`
	Record_Audio_Device    string   `json:"record_audio_device"`
	Record_Audio_Options   []string `json:"record_audio_options"`
	Record_Video_Device    string   `json:"record_video_device"`
	Record_Video_Options   []string `json:"record_video_options"`
	Record_Video_Advanced  []string `json:"record_video_advanced"`
	Record_Inspect_Models  []string `json:"record_inspect_models"`
	Has_Models             bool
	Record_Inspect_Backlog int     `json:"record_inspect_backlog"`
	Record_Duration        string  `json:"record_duration"`
	Train_Target           float64 `json:"train_target"`
	Train_Rate             float64 `json:"train_rate"`
	Train_Momentum         float64 `json:"train_momentum"`
	Train_Dropout          float64 `json:"train_dropout"`
}

// Map environment variables to Runtime
var Environment_Configation_Map = map[string]string{
	"DTRACK_WORKSPACE":              "Workspace",
	"DTRACK_WORKSPACE_KEEP_TEMP":    "Workspace_Keep_Temp",
	"DTRACK_RECORD_AUDIO_DEVICE":    "Record_Audio_Device",
	"DTRACK_RECORD_AUDIO_OPTIONS":   "Record_Audio_Options",
	"DTRACK_RECORD_VIDEO_DEVICE":    "Record_Video_Device",
	"DTRACK_RECORD_VIDEO_OPTIONS":   "Record_Video_Options",
	"DTRACK_RECORD_VIDEO_ADVANCED":  "Record_Video_Advanced",
	"DTRACK_RECORD_INSPECT_MODELS":  "Record_Inspect_Models",
	"DTRACK_RECORD_INSPECT_BACKLOG": "Record_Inspect_Backlog",
	"DTRACK_RECORD_DURATION":        "Record_Duration",
	"DTRACK_TRAIN_TARGET":           "Train_Target",
	"DTRACK_TRAIN_RATE":             "Train_Rate",
	"DTRACK_TRAIN_MOMENTUM":         "Train_Momentum",
	"DTRACK_TRAIN_DROUPOUT":         "Train_Dropout",
}

// Print information about configution file
func Show_Help() {
	fmt.Println("\nConfiguration Options:")
	fmt.Println("  Config.JSON Key\t\tEnvironment Variable\t\tDefault Value")
	fmt.Println("  ---------------\t\t--------------------\t\t-------------")
	fmt.Println("  workspace\t\t\tDTRACK_WORKSPACE\t\t_workspace")
	fmt.Println("  workspace_keep_temp\tDTRACK_WORKSPACE_KEEP_TEMP\tfalse")
	fmt.Println("  record_audio_device\t\tDTRACK_RECORD_AUDIO_DEVICE\tplughw")
	fmt.Println("  record_audio_options\t\tDTRACK_RECORD_AUDIO_OPTIONS\t[\"-f\", \"alsa\"]")
	fmt.Println("  record_video_device\t\tDTRACK_RECORD_VIDEO_DEVICE\t/dev/video0")
	fmt.Print("  record_video_options\t\tDTRACK_RECORD_VIDEO_OPTIONS")
	fmt.Println("\t[\"-f\", \"v4l2\", \"-framerate\", \"5\"]")
	fmt.Println("  record_video_advanced\t\tDTRACK_RECORD_VIDEO_ADVANCED\tSee Documentation")
	fmt.Println("  record_duration\t\tDTRACK_RECORD_DURATION\t\t00:10:00  (10 minutes)")
	fmt.Println("  record_inspect_models\t\tDTRACK_RECORD_INSPECT_MODELS\t[]")
	fmt.Println("  record_inspect_backlog\tDTRACK_RECORD_INSPECT_BACKLOG\t5")
	fmt.Println("  record_inspect_segment\tDTRACK_RECORD_INSPECT_SEGMENT\t-1")
	fmt.Println("  train_target\t\t\tDTRACK_TRAIN_TARGET\t\t0.95")
	fmt.Println("  train_rate\t\t\tDTRACK_TRAIN_RATE\t\t0.001")
	fmt.Println("  train_momentum\t\tDTRACK_TRAIN_MOMENTUM\t\t0.9")
	fmt.Println("  train_dropout\t\t\tDTRACK_TRAIN_DROUPOUT\t\t0.2")
	fmt.Println("\nOption Priority:")
	fmt.Println("  Defaults -> Config.json -> Environment")
}

// Loads Runtime configuration data into current state
func Load_Configuration(config_path string) {
	// Default configuration values
	cfg := Application_Configuration{
		Workspace:            "_workspace",
		Workspace_Keep_Temp:  false,
		Record_Audio_Device:  "plughw",
		Record_Audio_Options: []string{"-f", "alsa"},
		Record_Video_Device:  "/dev/video0",
		Record_Video_Options: []string{
			"-f", "v4l2", "-framerate", "15"},
		Record_Video_Advanced: []string{
			"-crf", "23", "-preset", "fast", "-maxrate", "3M", "-bufsize", "24M",
			"-tune", "zerolatency", "-filter_complex", "[1:v]drawtext" +
				"=fontfile=/usr/share/fonts/truetype/freefont/FreeMonoBold.ttf" +
				":text=%{localtime}:fontcolor=red@0.9:x=7:y=7:fontsize=48[dtstamp]"},
		Record_Duration:        "00:10:00",
		Record_Inspect_Models:  []string{},
		Record_Inspect_Backlog: 5,
		Train_Target:           0.95,
		Train_Rate:             0.001,
		Train_Momentum:         0.9,
		Train_Dropout:          0.2,
	}

	// Check for configuration file
	log.Debug("Loading configuration values from: %s", config_path)
	if _, err := os.Stat(config_path); err != nil {
		log.Info("Configuration file not found; using defaults.")
		Runtime = cfg
		return
	}

	// Load configuration file
	file_data, err := os.ReadFile(config_path)
	if err != nil {
		log.Die("Error opening configuration file; ABORT!")
	}

	// Merge configuration values into cfg
	if err := json.Unmarshal(file_data, &cfg); err != nil {
		log.Die("Failed to parse configuration as JSON; ABORT!")
	}

	// Search for known environment variables
	for env_key, conf_field := range Environment_Configation_Map {
		if env_value := os.Getenv(env_key); env_value != "" {
			log.Debug("Environment variable found: %s", env_key)
			field := reflect.ValueOf(&cfg).Elem().FieldByName(conf_field)
			if !field.IsValid() || !field.CanSet() {
				log.Die("Invalid field: %s", conf_field)
			}

			// Merge environment variables into cfg
			switch field.Kind() {
			case reflect.String:
				field.SetString(env_value)
			case reflect.Int:
				if intVal, err := strconv.Atoi(env_value); err == nil {
					field.SetInt(int64(intVal))
				} else {
					log.Die("%s is not Integer", env_key)
				}
			case reflect.Float64:
				if intVal, err := strconv.Atoi(env_value); err == nil {
					field.SetFloat(float64(intVal))
				} else {
					log.Die("%s is not Float64", env_key)
				}
			case reflect.Bool:
				if boolVal, err := strconv.ParseBool(env_value); err == nil {
					field.SetBool(boolVal)
				} else {
					log.Die("%s is not Boolean", env_key)
				}
			default:
				log.Die("Unexpected field type for %s", conf_field)
			}
		}
	}

	// Helper variables
	cfg.Has_Models = len(cfg.Record_Inspect_Models) > 0

	// Update session variables
	Runtime = cfg
}
